qa/statsd/src/cases/09.py

```diff
@@ -0,0 +1,81 @@
+#!/usr/bin/env pmpython
+# -*- coding: utf-8 -*-
+
+# Exercises hdr histogram aggregation on duration metrics
+
+import sys
+import socket
+import glob
+import os
+
+utils_path = os.path.abspath(os.path.join("utils"))
+sys.path.append(utils_path)
+
+import pmdastatsd_test_utils as utils
+
+utils.print_test_file_separator()
+print(os.path.basename(__file__))
+
+ip = "0.0.0.0"
+port = 8125
+sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+
+expected_min = 0
+expected_max = 20000000
+expected_count = 10000000
+expected_average = 10000000
+expected_median = 10000000
+expected_percentile90 = 18000000
+expected_percentile95 = 19000000
+expected_percentile99 = 19800000
+expected_stddev = 5773500.278068
+
+# so that we can guarantee same duration values which may vary when histogram aggregation is used - this is not the point of this test
+hdr_duration_aggregation = os.path.join("configs", "single", "duration_aggregation_type", "1", "pmdastatsd.ini")
+
+def run_test():
+    utils.print_test_section_separator()
+    utils.pmdastatsd_install(hdr_duration_aggregation)
+    for i in range(0, 10000001):
+        sock.sendto("test_hdr:{}|ms".format(i * 2).encode("utf-8"), (ip, port))
+    labels_output = utils.request_metric("statsd.test_hdr")
+    output = utils.get_instances(labels_output)
+    for k, v in output.items():
+        status = False
+        number_value = float(v)
+        if k == "/average":
+            if utils.check_is_in_bounds(expected_average, number_value):
+                status = True
+        elif k == "/count":
+            # TODO: Ask Nathan, if this is OK
```

Comment:
Which part?  The 0.5?

Reply:
Yes, that one.

Reply:
Yes, that's fine - it's a common pattern - when writing shell tests we even have a shared _within_tolerance() function available for this purpose.

Reply:
Alright, I will leave it as is and remove the TODO comment, thanks.

Reply:
I had to set up a new VM to work on this on which I encounter following phenomena: even 0.5 toleration margin is not enough, particularly in relation to count of received UDP payloads, even after setting ```sysctl net.core.rmem_max``` and ```sysctl net.core.rmem_default``` to 26214400, which makes me worry about reliability of such a test - even though its UDP after all, what would you approach to this be?

Reply:
Set the acceptable boundaries sufficiently wide that any reasonable scenario would be catered for.  Add comments to the test discussing possible causes for values outside these boundaries, in case people have to come along later and widen the boundaries further.  Even with wide boundaries, its often still likely to be a worthwhile test.

Reply:
Will do, thanks.